components/ws-manager-api/typescript/src/constraints.ts

```diff
@@ -95,19 +91,16 @@ export function constraintInverseHasPermissions(...permissions: PermissionName[]
  * @returns
  */
 export function constraintHasPermissions(...permissions: PermissionName[]): Constraint {
-    return (all: WorkspaceClusterWoTLS[], user: ExtendedUser, workspace: Workspace, instance: WorkspaceInstance) => {
+    return (all: WorkspaceClusterWoTLS[], args: ConstraintArgs) => {
         return all.filter((cluster) => permissions.some((p) => hasPermissionConstraint(cluster, p)));
     };
 }
 
-export function constraintRegional(
-    all: WorkspaceClusterWoTLS[],
-    user: ExtendedUser,
-    workspace: Workspace,
-    instance: WorkspaceInstance,
-): WorkspaceClusterWoTLS[] {
-    // TODO(cw): implement me
-    return [];
+export function constraintRegional(all: WorkspaceClusterWoTLS[], args: ConstraintArgs): WorkspaceClusterWoTLS[] {
+    if (!args.region || args.region.length === 0) {
+        return [];
```

Comment:
Shouldn't this return the "full list" of regions rather than empty? If we do not specify a region, the constraint should be set to "constrain to all clusters" rather than no cluster. This is because if I specify a region "unknown", the constraints should return the empty list, and the scheduling should fail.

Reply:
Alternatively, it's not a constraint and instead it's a preference, which may, or may not be honoured depending on capacity, etc.

Reply:
There will be multiple sets being returned (see above `workspaceClusterSets`) and the regional one would be empty, so the client resorts to the next level. I didn't want to change too much in this entire logic but I believe it could be simplified to something like filter and sort.